### PR TITLE
Provider side decoder API documentation

### LINKED ISSUE #13949: Decoder/Encoder documentation
This pertains to OpenSSL3 encoder/decoder documentation (feel free to tick checkboxes when reviewing issue(s)).

- [x] Wrong API description in [OSSL_ENCODER](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_ENCODER.pod) and [OSSL_DECODER](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_DECODER.pod): "OSSL_DECODER_do_all_provided() [...] data as arguments.": *data* is no argument to the function, *arg* is.
- [x] Wrong wording in [OSSL_DECODER](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_DECODER.pod): "OSSL_DECODER_do_all_provided() traverses all encoder implementations by all activated providers" should refer to *decoders*.
- [ ] Missing: provider-decoder (analogous to [provider-encoder](https://github.com/openssl/openssl/blob/master/doc/man7/provider-encoder.pod))
- [x] Nit: Wrong HTML markup, e.g., https://www.openssl.org/docs/manmaster/man7/ "$description$lt;-$description$gt;"

### PR DESCRIPTION
**Pull request by t8m**
Fixes #13949

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Thanks for doing that!

File: doc/man7/provider-encoder.pod [hunk 1]
```diff
@@ -152,6 +152,14 @@ An implementation with that output type outputs PEM formatted data.
 
 An implementation with that output type outputs DER formatted data.
 
+=item msblob
```
Comment (slontis) [inline doc/man7/provider-encoder.pod:155]: Not sure that it matters but these items are all uppercase in the code.

File: doc/man7/provider-decoder.pod [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+=head1 DESCRIPTION
+
+I<We use the wide term "decode" in this manual.  This includes but is
+not limited to deserialization.>
+
+The DECODER operation is a generic method to create a provider-native
+object reference or intermediate decoded data from an encoded form
+read from the given B<OSSL_CORE_BIO>. If the caller wants to decode
+a memory data, it should provide a L<BIO_s_mem(3)> B<BIO>. The decoded
```
Comment (slontis) [inline doc/man7/provider-decoder.pod:50]: decode data to memory

File: doc/man7/provider-decoder.pod [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+ /* Functions to export a decoded object */
+ void *OSSL_FUNC_decoder_export_object(void *ctx,
+                                       const void *objref, size_t objref_sz,
+                                       OSSL_CALLBACK *export_cb,
+                                       void *export_cbarg);
+
+=head1 DESCRIPTION
+
+I<We use the wide term "decode" in this manual.  This includes but is
```
Comment (slontis) [inline doc/man7/provider-decoder.pod:44]: The term "decode" is used throughout this manual..

File: doc/man7/provider-decoder.pod [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+pointer than being able to pass it to the appropriate BIO upcalls (see
+L<provider-base(7)/Core functions>).
+
+The DECODER implementation may be part of a chain, where data is
+passed from one to the next.  For example, there may be an
+implementation to decode an object from PEM to DER, and another one
+that decodes DER to an provider-native object.
+
+As the last decoding step in the decoding chain is supposed to create
+a provider-native object referenced by an object reference the
```
Comment (slontis) [inline doc/man7/provider-decoder.pod:64]: comma?

File: doc/man7/provider-decoder.pod [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+=item pvk
+
+An implementation with that input type decodes PVK formatted data.
+
+=back
+
+=item structure
+
+This property is used to specify the structure that the decoded data is
```
Comment (slontis) [inline doc/man7/provider-decoder.pod:144]: should we list all the internal types here (from top of decoders.inc)..

File: doc/man7/provider-encoder.pod (hunk 1 above)
Comment (t8m) [reply doc/man7/provider-encoder.pod:155]: Not really. The properties are specified as macro parameters in encoders.inc as `ENCODER("RSA", rsa, yes, msblob),` As you can see the msblob is lowercase.

File: doc/man7/provider-decoder.pod (hunk 2 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod:50]: `data from memory` but yes.

File: doc/man7/provider-decoder.pod (hunk 3 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod:44]: OK

File: doc/man7/provider-decoder.pod (hunk 4 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod:64]: I've reworded it.

File: doc/man7/provider-decoder.pod (hunk 5 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod:143]: done

**Comment by t8m** [comment]
Fixups pushed. PTAL

File: doc/man7/provider-decoder.pod (hunk 3 above)
Comment (richsalz) [reply doc/man7/provider-decoder.pod]: Don't say "we"  For example, in `d2i_X509` we (sic) say "In the decription here," so like "In the description here, the term `decode` includes ..."  Similar construct use in `DEFINE_STACK_OF.pod` and almost all "generic" manpages.

File: doc/man7/provider-decoder.pod [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+  */
+
+ /* Decoder parameter accessor and descriptor */
+ const OSSL_PARAM *OSSL_FUNC_decoder_gettable_params(void *provctx);
+ int OSSL_FUNC_decoder_get_params(OSSL_PARAM params[]);
+
+ /* Functions to construct / destruct / manipulate the decoder context */
+ void *OSSL_FUNC_decoder_newctx(void *provctx);
+ void OSSL_FUNC_decoder_freectx(void *ctx);
+ const OSSL_PARAM *OSSL_FUNC_decoder_settable_ctx_params(void *provctx)
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:24]: Missing ;

File: doc/man7/provider-decoder.pod [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+ /* Functions to export a decoded object */
+ void *OSSL_FUNC_decoder_export_object(void *ctx,
+                                       const void *objref, size_t objref_sz,
+                                       OSSL_CALLBACK *export_cb,
+                                       void *export_cbarg);
+
+=head1 DESCRIPTION
+
+I<The term "decode" is used throughout this manual.  This includes but is
+not limited to deserialization.>
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:45]: Not sure I understand the distinction. We have used the two terms synonymously.

File: doc/man7/provider-decoder.pod [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+to the I<metadata_cb> as B<OSSL_PARAM> parameters.
+
+The decoder doesn't need to know more about the B<OSSL_CORE_BIO>
+pointer than being able to pass it to the appropriate BIO upcalls (see
+L<provider-base(7)/Core functions>).
+
+The DECODER implementation may be part of a chain, where data is
+passed from one to the next.  For example, there may be an
+implementation to decode an object from PEM to DER, and another one
+that decodes DER to an provider-native object.
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:61]: an -> a

File: doc/man7/provider-decoder.pod [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+All "functions" mentioned here are passed as function pointers between
+F<libcrypto> and the provider in B<OSSL_DISPATCH> arrays via
+B<OSSL_ALGORITHM> arrays that are returned by the provider's
+provider_query_operation() function
+(see L<provider-base(7)/Provider Functions>).
+
+All these "functions" have a corresponding function type definition
+named B<OSSL_FUNC_{name}_fn>, and a helper function to retrieve the
+function pointer from a B<OSSL_DISPATCH> element named
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:76]: a -> an

File: doc/man7/provider-decoder.pod [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+from PEM input should be named "DER".
+
+Properties can be used to further specify details about an implementation:
+
+=over 4
+
+=item input
+
+This property is used to specify what format of input the implementation
+can decode.  Currently known input types are:
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:119]: Perhaps this should say something like "OpenSSL providers use the following input types", or similar. The way it is currently written seems to imply to me that provider authors are restricted to this set - which is not the case.

File: doc/man7/provider-decoder.pod [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+
+This property is used to specify the structure that the decoded data is
+expected to have. An example could be C<pkcs8>, to specify explicitly that
+the object to be decoded (presumably an asymmetric key pair, in this case)
+is wrapped in a PKCS#8 structure.
+
+=back
+
+The possible values of both these properties is open ended.  A provider may
+very well specify output types and structures that libcrypto doesn't know
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:151]: We talk about "output types" here - but that hasn't been used before. Previously we've talked about "input" types.

File: doc/man7/provider-decoder.pod [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+anything about.
+
+=head2 Subset selections
+
+Sometimes, an object has more than one subset of data that is interesting to
+treat separately or together.  It's possible to specify what subsets are to
+be decoded, with a set of bits I<selection> that are passed in an B<int>.
```
Comment (mattcaswell) [inline doc/man7/provider-decoder.pod:158]: An example might be good here, e.g. parameters, or public key.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man7/provider-decoder.pod (hunk 7 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod:45]: I copied the sentence from provider-encoder.pod. IMO serialization/deserialization has more specific meaning of recording/loading complete state of an object. Here for example an pem to der decoder does decoding some data from one form to another but I would not call that operation deserialization. On the other hand the whole operation of decoding PEM private key data into an EVP_PKEY can surely be called deserialization of the EVP_PKEY object.

Comment (t8m) [reply doc/man7/provider-decoder.pod:45]: I'll add `...  as individual decoders can do also decoding into intermediate data formats.`

File: doc/man7/provider-decoder.pod (hunk 6 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod:24]: OK

**Comment by t8m** [comment]
@mattcaswell fixups addressing your comments pushed.

File: doc/man7/provider-decoder.pod [hunk 13]
```diff
@@ -41,13 +41,14 @@ provider-decoder - The OSSL_DECODER library E<lt>-E<gt> provider functions
 
 =head1 DESCRIPTION
 
-I<We use the wide term "decode" in this manual.  This includes but is
-not limited to deserialization.>
+I<The term "decode" is used throughout this manual.  This includes but is
+not limited to deserialization as individual decoders can do also
```
Comment (slontis) [inline doc/man7/provider-decoder.pod:45]: also do

**Review by slontis — APPROVED** [review]
Apart from a small NIT this LGTM.

File: doc/man7/provider-decoder.pod (hunk 13 above)
Comment (t8m) [reply doc/man7/provider-decoder.pod]: Fixup pushed. I assume no re-approval is needed.

Comment (slontis) [reply doc/man7/provider-decoder.pod]: I must be tired - it took me a few gos to put the emoji in the correct place..

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews and comments.

**Pull request closed without merging**